q2_assembly/quast/tests/test_quast.py

```diff
@@ -791,15 +794,246 @@ def _copy(dst_dir):
                 make_artifact=make_artifact,
             )
 
-            tab_report, _ = evaluate_contigs(ctx=mock_ctx, contigs=contigs)
-            tab_report.validate()
+            refs_dir = self.get_data_path("references")
+            _copy_references(refs_dir, tmp)
 
+            refs = Artifact.import_data("GenomeData[DNASequence]", refs_dir)
+            tab_report, _, _ = evaluate_contigs(
+                ctx=mock_ctx, contigs=contigs, references=refs
+            )
             make_artifact.assert_called_once_with(
                 "QUASTResults", self.assert_is_dataframe((2, 43))
             )
+
+        tab_report.validate()
+        self.assertEqual(action.call_args[0][0], contigs)
+        export_data.assert_called_once_with(os.path.join(tmp, "vis_files"))
+
+    def test_evaluate_contigs_pipeline_no_refs_normal(self):
+        # this is used for mocking
+        def _copy(dst_dir):
+            os.makedirs(dst_dir, exist_ok=True)
+            shutil.copy2(
+                src=os.path.join(
+                    self.get_data_path("quast-results"), "enhanced_tabular_results.tsv"
+                ),
+                dst=os.path.join(dst_dir, "quast_results.tsv"),
+            )
+
+        contigs = Artifact.import_data(
+            "SampleData[Contigs]", self.get_data_path("contigs")
+        )
+
+        with tempfile.TemporaryDirectory() as tmp, patch(
+            "tempfile.TemporaryDirectory"
+        ) as mock_temp_dir:
+            mock_temp_dir.return_value.__enter__.return_value = tmp
+            export_data = MagicMock(
+                side_effect=lambda x: _copy(os.path.join(x, "quast_data"))
+            )
+            action = MagicMock(return_value=(MagicMock(export_data=export_data),))
+            ctx = Context()
+            ctx._scope = MagicMock()
+            make_artifact = MagicMock(
+                side_effect=lambda *args: ctx.make_artifact(*args)
+            )
+            mock_ctx = MagicMock(
+                spec=Context,
+                get_action=MagicMock(return_value=action),
+                make_artifact=make_artifact,
+            )
```

Comment:
You could also write a small method to generate the mocked context and use it in other tests.